perf/src/main.rs

```diff
@@ -0,0 +1,118 @@
+    let key = quinn::PrivateKey::from_der(&cert.serialize_private_key_der()).unwrap();
+    let cert = quinn::Certificate::from_der(&cert.serialize_der().unwrap()).unwrap();
+
+    let mut server_config = quinn::ServerConfigBuilder::default();
+    server_config
+        .certificate(quinn::CertificateChain::from_certs(vec![cert.clone()]), key)
+        .unwrap();
+    server_config.protocols(&[b"perf"]);
+
+    let server_config = server_config.build();
+
+    let mut endpoint = quinn::EndpointBuilder::default();
+    endpoint.listen(server_config);
+
+    let (endpoint, mut incoming) = endpoint
+        .bind(&SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0))
+        .unwrap();
+
+    info!("listening on {}", endpoint.local_addr().unwrap());
+
+    while let Some(handshake) = incoming.next().await {
+        tokio::spawn(async move {
+            if let Err(e) = accept(handshake).await {
+                error!("connection lost: {}", e);
+            }
+        });
+    }
+}
+
+async fn accept(handshake: quinn::Connecting) -> Result<()> {
+    let quinn::NewConnection {
+        uni_streams,
+        bi_streams,
+        connection,
+        ..
+    } = handshake.await.context("handshake failed")?;
+    debug!("{} connected", connection.remote_address());
+    tokio::select! {
+        x = drive_uni(connection, uni_streams) => x?,
+        x = drive_bi(bi_streams) => x?,
+    }
+    Ok(())
+}
+
+async fn drive_uni(
+    connection: quinn::Connection,
+    mut streams: quinn::IncomingUniStreams,
+) -> Result<()> {
+    while let Some(stream) = streams.next().await {
+        let stream = stream?;
+        let connection = connection.clone();
+        tokio::spawn(async move {
+            if let Err(e) = accept_uni(connection, stream).await {
+                error!("request failed: {}", e);
+            }
+        });
+    }
+    Ok(())
+}
+
+async fn accept_uni(connection: quinn::Connection, stream: quinn::RecvStream) -> Result<()> {
+    let bytes = read_req(stream).await?;
+    let response = connection.open_uni().await?;
+    respond(bytes, response).await?;
+    Ok(())
+}
+
+async fn drive_bi(mut streams: quinn::IncomingBiStreams) -> Result<()> {
+    while let Some(stream) = streams.next().await {
+        let (send, recv) = stream?;
+        tokio::spawn(async move {
+            if let Err(e) = accept_bi(send, recv).await {
+                error!("request failed: {}", e);
+            }
+        });
+    }
+    Ok(())
+}
+
+async fn accept_bi(send: quinn::SendStream, recv: quinn::RecvStream) -> Result<()> {
+    let bytes = read_req(recv).await?;
+    respond(bytes, send).await?;
+    Ok(())
+}
+
+async fn read_req(mut stream: quinn::RecvStream) -> Result<u64> {
+    let mut buf = [0; 8];
+    stream
+        .read_exact(&mut buf)
+        .await
+        .context("reading request")?;
+    let n = u64::from_be_bytes(buf);
+    debug!("got req for {} bytes", n);
```

Comment:
This actually misses draining the remaining request body, which is important for upload tests.

I would recommend to copy&paste https://github.com/quinn-rs/quinn/blob/main/bench/src/bulk.rs#L88-L101.
Optionally a `async fn drain()` which does the same could also be added to `RecvStream`.